src/adapter/src/coord/read_policy.rs

```diff
@@ -783,5 +785,11 @@ impl crate::coord::Coordinator {
                     .unwrap_or_terminate("cannot fail to set read policy");
             }
         }
+
+        for mut read_holds in read_holdses {
+            // Make sure the Drop impl does not send this along for releasing!
+            read_holds.dropped_read_holds_tx.take();
+            read_holds.holds.clear();
+        }
```

Comment:
Related to another one of my comments, but if this method just had `read_holdses: Vec<HashMap<Antichain<Timestamp>, CollectionIdBundle>>` as input, then I don't think we'd need to worry about this.

Reply:
I wasn't sure about changing this signature. I'm leaving in place a bunch of explicit `release_read_holds()` calls that this would break.

I _could_ of course remove all of those and purely rely on the Drop impl. What would you prefer?

Reply:
It's changed in the PR now, because keeping the explicit `release_read_holds()` is tricky. There's now a split between `InternalReadHold` and `ReadHold`, which makes something explicit that has been there for a while in the adapter: there's `ReadHolds` that we create when creating a collection, and we keep those around forever and regularly call `update_read_hold())` (which downgrades it, naturally!), and there's `ReadHolds` that we acquire for peeks and such.

With this PR, the former is now `InternalReadHold` and the latter is just `ReadHold` (which is droppable and wraps a `InternalReadHold`.

The split also shows up in the fact that we have two different places where we create `ReadHolds`:

1. for short-lived ones: https://github.com/MaterializeInc/materialize/blob/eca245f1e55bd9fd7c4bda9297107548444563d1/src/adapter/src/coord/read_policy.rs#L553
2. for the long-lived ones: https://github.com/MaterializeInc/materialize/blob/eca245f1e55bd9fd7c4bda9297107548444563d1/src/adapter/src/coord/read_policy.rs#L294

The code path for the long-lived holds also does some initialization, but then it does basically the same thing as `acquire_read_holds`. I didn't want to mess with this too much in this PR, though.